src/utils/latch.ts

```diff
@@ -1,25 +1,28 @@
-import timers from 'timers';
+/**
+ * Interface Latch is a simple extension on Promise that is resolved via calling
+ * a method.  It is essentially a simplified barrier.
+ *
+ * @see https://en.wikipedia.org/wiki/Barrier_(computer_science)
+ */
+interface Latch extends Promise<void> {
+  /** Calling the resolve() method resolves the Latch. */
+  resolve(): void;
+}
 
 /**
- * class Latch is a simple extension on Promise that is resolved via calling a
- * method.
+ * Creates a Latch that is an extension of a Promise that can be resolved via
+ * calling a method on that Promise.
  */
-export default class Latch extends Promise<void> {
-  #resolve?: () => void;
-  constructor() {
-    super((resolve) => {
-      // We can't set the property from within the callback in the superclass,
-      // because our instance hasn't been constructed yet.  So we need to do it
-      // in a setImmediate() callback.
-      timers.setImmediate(() => {
-        this.#resolve = resolve;
-      });
-    });
-  }
+export default function Latch(): Latch {
+  const holder: {resolve?: () => void} = {};
+  const result: Latch = new Promise<void>((resolve) => {
+    holder.resolve = resolve;
+  }) as any;
 
-  resolve() {
-    if (this.#resolve) {
-      this.#resolve();
-    }
+  if (!holder.resolve) {
```

Comment:
From my reading of https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise this line should execute before the Promise's `resolve` argument is called. How can this test be true?

Reply:
Looking at the [spec](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-promise-executor), I believe the call to the executor is always synchronous? That is, it _always_ calls the function at step 9 (which _typically_ sets up a callback) immediately.  It then returns the promise at step 11, after the call has completed.